homeassistant/components/climate/spider.py

```diff
@@ -0,0 +1,140 @@
+"""
+Support for Spider thermostats.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/climate.spider/
+"""
+
+import logging
+
+from homeassistant.components.climate import (
+    ATTR_TEMPERATURE, STATE_COOL, STATE_HEAT, STATE_IDLE,
+    SUPPORT_OPERATION_MODE, SUPPORT_TARGET_TEMPERATURE, ClimateDevice)
+from homeassistant.components.spider import DOMAIN as SPIDER_DOMAIN
+from homeassistant.const import TEMP_CELSIUS
+
+DEPENDENCIES = ['spider']
+
+OPERATION_LIST = [
+    STATE_HEAT,
+    STATE_COOL,
+]
+
+HA_STATE_TO_SPIDER = {
+    STATE_COOL: 'Cool',
+    STATE_HEAT: 'Heat',
+    STATE_IDLE: 'Idle'
+}
+
+SPIDER_STATE_TO_HA = {value: key for key, value in HA_STATE_TO_SPIDER.items()}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Spider thermostat."""
+    if discovery_info is None:
+        return
+
+    devices = [SpiderThermostat(hass.data[SPIDER_DOMAIN]['controller'], device)
+               for device in hass.data[SPIDER_DOMAIN]['thermostats']]
+    add_devices(devices, True)
+
+
+class SpiderThermostat(ClimateDevice):
+    """Representation of a thermostat."""
+
+    def __init__(self, api, thermostat):
+        """Initialize the thermostat."""
+        self.api = api
+        self.thermostat = thermostat
+        self.master = self.thermostat.has_operation_mode
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        supports = SUPPORT_TARGET_TEMPERATURE
+
+        if self.thermostat.has_operation_mode:
+            supports = supports | SUPPORT_OPERATION_MODE
+
+        return supports
+
+    @property
+    def unique_id(self):
+        """Return the id of the thermostat, if any."""
+        return self.thermostat.id
+
+    @property
+    def name(self):
+        """Return the name of the thermostat, if any."""
+        return self.thermostat.name
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        return TEMP_CELSIUS
+
+    @property
+    def current_temperature(self):
+        """Return the current temperature."""
+        return self.thermostat.current_temperature
```

Comment:
Does this attribute (or other) access involve I/O or is it just returning a cached value? I/O is not permitted in entity properties, since they are evaluated in the event loop.

Reply:
the object "thermostat" is cached within this class itself. It depends on this object in the library;
https://github.com/peternijssen/spiderpy/blob/master/spiderpy/devices/thermostat.py

I guess because of the for loop within that class, I should cache it either in HA or the library?

Reply:
That looks fine. It's just parsing a json dict. Great!

Reply:
You could add some caching in the library to avoid unnecessary for loops, but it's not required.

Reply:
Agree. I will take care of that in the next version.